src/gateway.c

```diff
@@ -288,20 +299,41 @@ static int handle_prepare(struct handle *req)
 		tracef("handle prepare sqlite prepare failed %d", rc);
 		stmt__registry_del(&g->stmts, stmt);
 		failure(req, rc, sqlite3_errmsg(g->leader->conn));
-		return 0;
+		return 0; // ignored
 	}
 
 	if (stmt->stmt == NULL) {
 		tracef("handle prepare empty statement");
 		stmt__registry_del(&g->stmts, stmt);
 		failure(req, rc, "empty statement");
-		return 0;
+		return 0; // ignored
 	}
 
 	response.db_id = (uint32_t)request.db_id;
 	response.id = (uint32_t)stmt->id;
 	response.params = (uint64_t)sqlite3_bind_parameter_count(stmt->stmt);
 	SUCCESS(stmt, STMT);
+	return 0; // ignored
+}
+
+static void prepareBarrierCb(struct barrier *barrier, int status)
+{
+	tracef("prepare barrier cb status:%d", status);
+	(void)prepareBarrierCbInner(barrier, status);
```

Comment:
Oh I see. Besides the fact that it feels a bit of a smell, it also means that actually the checks performed by macros like `START` will be effectively ignored no?

For example if `rv_ = request_##REQ##__decode(cursor, &request);` returns non-0 then `prepareBarrierCbInner` will silently return without providing any response to the client, and effectively hanging the request forever. Unless I'm missing something.

It'd be nice to keep using those macros at the very beginning of the request processing chain, possibly saving on the `gateway` structure any data that you need to propagate (like it's done with `gateway->sql`). Or something along these lines, I didn't really think a lot about this, just pointing that this trick feels/is probably suboptimal.

Reply:
> For example if rv_ = request_##REQ##__decode(cursor, &request); returns non-0 then prepareBarrierCbInner will silently return without providing any response to the client, and effectively hanging the request forever. Unless I'm missing something.

Yes, that's a good point. I thought this was the same as the current behavior, but it's not -- currently, if (say) `START` fails in `handle_prepare` we end up closing the gateway/connection:

https://github.com/canonical/dqlite/blob/3a9bb7abc316507ad77e80a9df22713dafb72731/src/conn.c#L150-L155

Which is maybe not ideal, but better than silently dropping the message.

> It'd be nice to keep using those macros at the very beginning of the request processing chain, possibly saving on the gateway structure any data that you need to propagate (like it's done with gateway->sql).

Yes, I avoided this on the first pass for the sake of keeping the change simple, but it would allow solving this problem in a less hacky way. I'll take a look at implementing it.

Reply:
> > For example if rv_ = request_##REQ##__decode(cursor, &request); returns non-0 then prepareBarrierCbInner will silently return without providing any response to the client, and effectively hanging the request forever. Unless I'm missing something.
> 
> Yes, that's a good point. I thought this was the same as the current behavior, but it's not -- currently, if (say) `START` fails in `handle_prepare` we end up closing the gateway/connection:
> 
> https://github.com/canonical/dqlite/blob/3a9bb7abc316507ad77e80a9df22713dafb72731/src/conn.c#L150-L155
> 
> Which is maybe not ideal, but better than silently dropping the message.

In this specific case closing the connection might be the best you can do since if decoding fails it means the client sent some invalid data and it would be difficult to recover the communication from that. It should basically happen only if a client is buggy.

> 
> > It'd be nice to keep using those macros at the very beginning of the request processing chain, possibly saving on the gateway structure any data that you need to propagate (like it's done with gateway->sql).
> 
> Yes, I avoided this on the first pass for the sake of keeping the change simple, but it would allow solving this problem in a less hacky way. I'll take a look at implementing it.

Perhaps this data should be added to the request object instead of the gateway object, but I'm just hand-waving, as I did really look at the details.

Reply:
Okay, the latest series of commits stores the new `uint64_t db_id` and `struct stmt *registered_stmt` fields in the handle, but continues to use existing gateway fields where they're available. I could do a separate refactoring PR to move some request-scoped gateway fields to the handle.